### PR TITLE
kanidm: add support for multiple versions

### LINKED ISSUE #353889: Package kanidm minor versions separately
### Describe the bug
Similar to PostgreSQL, Kanidm has a manual upgrade process to move between minor versions (e.g. 1.3 -> 1.4), and can only upgrade one minor version at a time. Hence, upgrading NixOS packages might invariably break Kanidm (especially on NixOS unstable). It would be great to be able to pin Kanidm to a minor version (i.e. services.kanidm.package = kanidm_1_4) to be able to avoid breakage.

### Steps To Reproduce
1. Update NixOS
2. Kanidm no longer starts because it was upgraded to a version that needs manual intervention

### Expected behavior
Like other databases, Kanidm should be explicitly pinned to a specific version to avoid this.

### Notify maintainers

@adamcstephens @Flakebi 

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.59, NixOS, 24.11 (Vicuna), 24.11.20241103.b47af37`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by adamcstephens**
Adds package support for multiple versions. ~~Will probably require updateScript modifictions, but putting that off for now.~~

Sets default in module to the version at NixOS release time. kanidm will release new, potentially breaking changes during each NixOS release, at which point users will need to manually update their configuration.

This will be manually backported, with the stateVersion updated for 24.05, to verify the workflow and ensure support for the remaining weeks of 24.05.

Resolves #353889
Closes #361458

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
This won’t meet the `by-name` rules. I suggest taking a look at `openjfx`, `ninja`, and `apple-sdk` for some examples of how you can make multi‐version packages work while keeping `nixpkgs-vet` happy.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -1913,6 +1913,9 @@ with pkgs;
 
   gopass-summon-provider = callPackage ../tools/security/gopass/summon.nix { };
 
+  kanidm_1_3 = callPackage ../by-name/ka/kanidm/1_3.nix { };
+  kanidm_1_4 = callPackage ../by-name/ka/kanidm/1_4.nix { };
+
```
Comment (oddlama) [inline pkgs/top-level/all-packages.nix:1918]: Can we move this to around line ~11700 (category 'servers')? We already have `kanidmWithSecretProvisioning` declared there

Comment (adamcstephens) [reply pkgs/top-level/all-packages.nix]: Done, I also added provisioning packages for the new versioning.

**Review by oddlama — APPROVED** [review]

**Comment by adamcstephens** [comment]
Yes, https://github.com/NixOS/nixpkgs/pull/364105

**Pull request merged**